terraform/aws/main.tf

```diff
@@ -0,0 +1,551 @@
+
+#this aws terraform utilizes karpenter.sh to create a kubernetes cluster with autoscaling
+
+terraform {
+  required_version = "~> 1.0"
+
+  backend "s3" {
+    bucket         = "graphistry-tf-state"
+    key            = "terraform.tfstate"
+    region         = "us-east-1"
+  }
+
+  required_providers {
+    aws = {
+      source  = "hashicorp/aws"
+      version = "~> 4.0"
+    }
+    helm = {
+      source  = "hashicorp/helm"
+      version = "~> 2.5.1"
+    }
+    kubectl = {
+      source  = "gavinbunney/kubectl"
+      version = "~> 1.14"
+    }
+    kubernetes = {
+      source  = "hashicorp/kubernetes"
+      version = "~> 2.14.0"
+    }
+  }
+}
+
+provider "aws" {
+
+  region = var.availability_zone_name
+}
+
+locals {
+  cluster_name = var.cluster_name
+
+  # Used to determine correct partition (i.e. - `aws`, `aws-gov`, `aws-cn`, etc.)
+  partition = data.aws_partition.current.partition
+}
+
+data "aws_partition" "current" {}
+
+resource "aws_eks_addon" "addons" {
+  depends_on = [
+   aws_iam_instance_profile.karpenter 
+  ]
+  for_each                 = { for addon in var.addons : addon.name => addon }
+  cluster_name             = module.eks.cluster_id
+  addon_name               = each.value.name
+  addon_version            = each.value.version
+  resolve_conflicts = "OVERWRITE"
+}
+
+resource "aws_iam_instance_profile" "karpenter" {
+  name = "KarpenterNodeInstanceProfile-${local.cluster_name}"
+  role = module.eks.eks_managed_node_groups["nodegroup"].iam_role_name
+}
+
+module "karpenter_irsa" {
+  source  = "terraform-aws-modules/iam/aws//modules/iam-role-for-service-accounts-eks"
+  version = "4.17.1"
+
+  role_name                          = "karpenter-controller-${local.cluster_name}"
+  attach_karpenter_controller_policy = true
+
+  karpenter_controller_cluster_id = module.eks.cluster_id
+  karpenter_controller_node_iam_role_arns = [
+    module.eks.eks_managed_node_groups["nodegroup"].iam_role_arn
+  ]
+
+  oidc_providers = {
+    ex = {
+      provider_arn               = module.eks.oidc_provider_arn
+      namespace_service_accounts = ["karpenter:karpenter"]
+    }
+  }
+}
+
+
+module "vpc" {
+  # https://registry.terraform.io/modules/terraform-aws-modules/vpc/aws/latest
+  source  = "terraform-aws-modules/vpc/aws"
+  version = "3.12.0"
+
+  name = local.cluster_name
+  cidr = "10.0.0.0/16"
+
+  azs             = ["us-east-1a", "us-east-1b", "us-east-1c"]
```

Comment:
I assume users can override?

Reply:
Yea can turn these into variables to be overridden